segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/StorageMetadata.java

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.metadata;
+
+import io.pravega.common.io.serialization.VersionedSerializer;
+import io.pravega.segmentstore.storage.mocks.MockStorageMetadata;
+import lombok.Data;
+
+import java.io.Serializable;
+
+/**
+ * Storage Metadata.
+ * All storage related metadata is stored in {@link ChunkMetadataStore} using set of key-value pairs.
+ * The String value returned by {@link StorageMetadata#getKey()} is used as key.
+ * Notable derived classes are {@link SegmentMetadata} and {@link ChunkMetadata} which form the core of metadata related
+ * to {@link io.pravega.segmentstore.storage.chunklayer.ChunkStorageManager} functionality.
+ */
+@Data
+public abstract class StorageMetadata implements Serializable {
+
+    /**
+     * Retrieves the key associated with the metadata.
+     *
+     * @return key.
+     */
+    public abstract String getKey();
+
+    /**
+     * Creates a deep copy of this instance.
+     *
+     * @return
+     */
+    public abstract StorageMetadata deepCopy();
```

Comment:
Do we really need to make deep copies of `StorageMetadata` objects and how often do you expect to create them?

Reply:
Because java is reference based language, two transactions can end up pointing to the same record and overwrite each others changes. Therefore for complete isolation we do need  completely separate copies.  

You are right - currently for both `SegmentMetadata `and `ChunkMetadata` don't have any embedded objects - they are flat objects with few fields of primitive data types (long/string/int). 
But I think the name `deepCopy `captures the intention. The alternative are - `copy`() or `clone`(). Although `clone `is perhaps close enough to intended meaning .

Deep copy is made whenever with in any transaction a record is read from the underlying store or buffer for the first time.